pennylane/ops/pulse/convenience_functions.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This file contains convenience functions for pulse programming."""
+
+import numpy as np
+
+has_jax = True
+try:
+    import jax.numpy as jnp
+except ImportError:
+    has_jax = False
+
+
+def pwc(t):
+    """Create a function that is piecewise-constant in time, based on the params for a TDHamiltonian.
+
+    Args:
+            t(Union[float, tuple(float, float)]: the total duration as a float, or the start and end time as floats.
+            index(int): the index at which the relevant parameter array is located in the overall ``params`` variable
+
+    Returns:
+            func: a function that can be passed the full ``params`` variable and ``t``, and will return the
+                    corresponding constant
+
+    **Example**
+
+    >>> t1, t2 = 1, 3
+    >>> f1 = pwc((t1, t2))
+
+    The resulting function ``f1`` has the call signature ``f1(params, t)``. If passed an array of parameters and
+    a time, it will assign the array as the constants in the piecewise function, and select the constant corresponding
+    to the specified time, based on the time interval defined by ``t``.
+
+    >>> params = [np.linspace(10, 20, 10)]
+    >>> f1(params, 2)
+    tensor(15.55555556, requires_grad=True)
+
+    >>> f1(params, 2.1)  # same bin
+    tensor(15.55555556, requires_grad=True)
+
+    >>> f1(params, 2.5)  # next bin
+    tensor(17.77777778, requires_grad=True)
+    """
+    if not has_jax:
+        raise ImportError(
+            "Module jax is required for any pulse-related convenience function. "
+            "You can install jax via: pip install jax"
+        )
+
+    if isinstance(t, tuple):
+        t1, t2 = t
+    else:
+        t1 = 0
+        t2 = t
+
+    def func(params, t):
+        num_bins = len(params)
+        # include 0 as an additional option for function output
+        params = jnp.array(list(params) + [0])
+
+        # get idx from timestamp, then set idx=0 if idx is out of bounds for the array
+        idx = num_bins / (t2 - t1) * (t - t1)
+        idx = jnp.where((idx >= 0) & (idx <= num_bins), jnp.array(idx, dtype=int), -1)
+
+        return params[idx]
+
+    return func
+
+
+def pwc_from_function(t, num_bins):
```

Comment:
Can we use both functionalities under the same name `pwc`?

Reply:
Yes we can, but I feel like they do different enough things that I'm not sure if its helpful or confusing.

`pwc` takes a `timespan`, and returns a callable with call signature `(params, t)`, where `params` is expected to be an array defining the number and value of the bins.

`pwc_from_function` takes a `timespan` and a `num_bins`, and returns a callable with call signature `(function)`, which if called *then* returns a callable that can take `(params, t)`, where `params` is whatever it was defined to be in the function.

I can make it so if you pass `num_bins`, it works as a decorator and creates a callable that expects to receive a function, and if you don't, it goes straight to returning a callable with signature `(params, t)`. I.e. you can do:
```
>>> f = pwc((1, 3))
```
and then pass an array as params later when calling `H(params, t)`, or

```
>>> @pwc((1, 3), 10)
... def f(p, t):
...     return p * t
```

and pass whatever `p` is in the decorated function when calling `H(params, t)`.

If that doesn't seem confusing, it's easy to add. My concern would be that people would try passing `num_bins` and expect that they could then pass an array with `len=num_bins`and `t`. Then you get a "takes 1 positional argument but 2 were given" error. 

Let me know what you think. Maybe there is a better way to combine them than to base the behaviour on the `num_bins` kwarg.

Reply:
So far I think we only have limited use for `pwc_from_function` in mind (any at all?) whereas `pwc` is crucial. So I wouldnt jeopardize making the latter confusing for the sake of unifying them. I agree with your point @lillian542 that they have very different behavior so I suggest leaving them separate.

Reply:
Yes, that could be a bit confusing. Let's keep them separate then.